README.md
# bos-missionbuilder
Web-based mission builder for Il-2: BoS

<img src="images/screenshot1.jpg">

Currently work-in-progress.

The web GUI generates an intermediate mission format that abstracts away a lot of the complexity of the BoS/RoF mission file format. These intermediate files are plain JSON documents which currently are stored server-side in an embedded NoSQL database (ElasticSearch).

This software can be run by individual users on their local machines with the ability to directly export .Mission and .eng files to the BoS missions folder.

License: MIT license
Frameworks Web: jQuery, Handlebars, Twitter Bootstrap
Frameworks Server: Spring Boot, ElasticSearch, Java 8


-- Presentation --

The early alpha BoS Simple Mission Builder (SMB) is a 3rd party standalone mission builder for creating simple single player missions for IL-2: Battle of Stalingrad

TLDR: Extremely unofficial early alpha Web browser based mission builder for creating simple missions. Download, unzip, configure and run. Probably extremely buggy.


The main objective of SMB is to provide a mean to create simple single player missions without having access to the to-be-released official Full Mission Builder, while providing more functionality than the current Quick Mission builder present in BoS today.

Main features:
* Fully zoomable/pannable map of Stalingrad (extracted from BoS game resources)
* Specification of date, time and mission name/briefing.
* Allows placing of air, ground and static object groups anywhere on the map
* Editing of basic group properties (size, type, altitude, speed, heading...)
* Air and ground units can be given waypoints they follow. Includes altitude and speed changes.
* Each group can contain up to 24 units. This may result in really bad framerates if your system cannot handle it.
* Each waypoint may be given a Command to be executed. Supported commands are:
** START - Start flight from ground. Should be first waypoint of air groups not starting in the air and placed very close to the starting position
** LAND - Land flight. Should be last waypoint.
** ATTACK_AREA - Will search for targets in the designated area, includes "radius" property.
** ATTACK_TARGET - Attack target specified from list of selectable groups. Note - make sure you select a target from the enemy side...
** COVER - Escort the group specified from list of selectable groups.
* Includes city, airfield, train station and bridges loaded from BoS /Template folder. Note that only objects within the mission area bounds are loaded. Loading all cities does work in-game but mission load times becomes very very long.
* Option to automatically generate some AAA assets on airfields having an adjacent LAND command on them.
* Briefing map will currently show waypoint path of own flight and enemy ground targets. Will add more fine-granular control of briefing map icons later.
* Exports .Mission and .eng files to %INSTALL_DIR%/data/Missions/webmissions folder

* Note: Due to some limitation in BoS, one MUST start BoS in Offline mode (just like Chir Front missions). This means unlocks cannot be used. Regrettably, it also seems as AI planes cannot be equipped with unlocks.
* Note: There's no absolute guarantee generated missions will work. If the BoS GUI just returns you directly to the Missions list after starting to show the Loading splash screen, something is invalid in the .Mission file. Also, if you for some reason edit the generated .eng file manually, it must be saved in UTF-16 Lower Endian encoding or the mission won't load.

SMB works by using a Web-based GUI to drag n drop units onto the map, setting waypoints (with optional commands on them) and then selecting the "Export to disk" meny option to write a .Mission and a .eng file to a %INSTALLDIR%/data/Mission/webmissions folder.

Missions may take a long time to load, especially for the first mission of the day. The RoF FMB manual states that missions loaded from the textual .Mission format takes a lot longer to load. Expect 1-2 minutes for the first mission since BoS seems to do some kind of roads generation thing on first mission load.
The mission generation performed by the SMB is based on looking at .Mission files generated by BoS tracks and the Rise of Flight FMB manual. There are many features missing from SMB that I assume the FMB will later support.

License: The SMB is Open Source software licensed under the MIT License. For full License details, please see: 
Disclaimer: I am not in any way affiliated with 1C or 777 and I do NOT have beta access to the BoS FMB or any other non-public information. 

Requirements:
Java 8 Runtime Environment installed. Earlier versions of Java are NOT supported since SMB utilizes a lot of the new API's introduced in Java 8.
SMB requires a copy of BoS to be installed on the users hard drive as it utilizes some of the files found under the /Template directory.
SMB web GUI is _only_ tested using recent versions of Google Chrome. No other browsers are currently supported.

Installation:
For those of you that have tried the Bosparser application, SMB installs in a more or less identical way. Java 8 is required. Install from Oracle. If you think you have Java installed, open a command shell and type 'java -version' without ' to see Java runtime version.
1. Download the install zip from Dropbox:
2. Unzip somewhere on your hard drive
3. Open the config/application.properties file. Change the property bos.data.directory to point at your BoS installations /data folder. This property is used to load files from /Template as well as saving your missions directly into the /data/Missions/webmissions folder
4. Start the application by executing bos-smb.bat file. Java may give you an alert about Network access, I think you can ignore it but otherwise click yes.
5. Open your Web Browser and type "localhost". The SMB application should load, displaying the BoS map. Note: If you are using bosparser (or any other web server software using port 80) you will need to modify the http.port setting in config/application.properties to some other port number and then type localhost:[portnumber] instead in your browser.


Known issues:
-- Extremely limited set of loadouts. Since I do not actually know internal loadout IDs of BoS I have to look at generated .Mission files and try to figure out which "payloadId" for a given plane that matches a selectable payload in the GUI. Also - it seems as AI planes cannot use loadouts that needs to be unlocked.
-- The object selection/dragging is a bit clunky, especially when one wants to drag the map as well. Working on that...
-- START/LAND commands "works", but since I havn't been able to figure out how to extract landing/start coordinates from within BoS, AI planes will start/land outside of runways in weird directions. I strongly suggest letting AI flights start in the air.
-- COVER command seems to work. However, I havn't figured out how to let an AI flight stop covering and continuing following their waypoints.
-- ATTACK_AREA works quite fine for AI flights, just remember to set the Radius to something quite large. I've noticed that Stukas like to make a ~10-15 km approach before actually starting their bombing run. Non Dive-bombers have a much more direct attack approach.
-- ATTACK_AREA has a default timeout of 20? minutes after which planes will continue on their previous route.
-- ATTACK_TARGET seems to work OK. After target is destroyed, planes will currently NOT resume their route.
-- Vehicle groups told to use Formation ROAD_COLUMN doesn't actually follow any roads. Not sure why yet, I'm explicitly giving them the appropriate Formation command.
-- Artillery fire - I havn't been able to get either mobile artillery or stationary artillery to fire even when trying to set area or target attack commands. Not sure if its a limitation in BoS or if my generated commands are faulty.
-- Line formations gets skewed if initial group heading is not 0, 90, 180, 270. (Working on that)


More stuff, not sure if it's bugs in BoS or in my missions.
-- Ground unit engagement AI seems a bit weird. Tanks never turn their turrets and as far as I have seen, they don't turn to engage enemies even when just a few degrees to either side. If directly in front of each other, tanks seem to fire on enemy tanks.
-- However, I have seen a USSR heavy machine gun successfully engaging Pz III tanks driving by, turning with them and firing until the tank blows up (mostly after hitting the PZ III rear.

I will continue adding features when I have the time, but I would be very happy for any assistence offered by the community. Everything from feedback, payloadIds, coding help (GUI is HTML/javascript/jQuery, Backend is Java 8 / Spring) etc.

<b>Some more screenshots </b>

Create Ground Group dialog<br/> <br/>
<img src="images/screenshot2.jpg">
<br/>
<br/>
Waypoint edit dialog<br/>
<img src="images/screenshot3.jpg">
<br/>
<br/>
In-game briefing with text and map markers<br/>
<img src="images/screenshot4.jpg">
<br/>
<br/>
In-game flying, AI group of FW-190A3 attacking GAZ-M column. Note own tanks in foreground and more enemy ground units in the background.
<img src="images/screenshot5.jpg">

changelog.txt
-- 2016-01-06 --
* Stalingrad Summer / Winter now selectable when creating or editing mission
* Weather Editor now has Cloud Types for summer and winter. If summer map chosen with winter clouds, light summer weather will be set automatically. Sim goes very blue otherwise.
* Added Battle of Moscow planes currently available: Mig3, P-40, I-16, Bf109E7, Bf110 and MC202 are now addable to missions.
* Fixed bug where flight group name from previous group would be present in textfield.
* Fixed bug where changed unit thumbnail path in BoS install caused path to become invalid making thumbnail images disappear.

-- 2015-05-18 --
* Ju-52 can now be added to missions. Note: Must not be player-controlled.
* Thanks to community member KpgQuop, we now have a more or less full set of static objects we can include in missions.
* Minor bug fixes.

-- 2015-03-04 --
* Added Flight Manager, which replaces the Skin Manager. Allows AI level to be set for individual planes as well as skins. Will also add Loadouts to this view later on. Please note that I have locked the currently selected player plane for now in order to not make things over-complicated syncing this view with the controls in the Edit Flight form.
* Added all ground vehicles present in the FME to SMB
* Added thumbnail images for all ground vehicles (should help people figure out what a "61K" or "BA10M" is). Note - uses the preview images from within the BoS data/graphics/preview folder so if your application.properties file does not point at the correct folder you won't see any images (and export etc won't work)
* Added "effects". There are five different smokes and fires to choose from. IMPORTANT: I recommend using these sparingly since you must figure out the exact altitude for the effect somehow. SMB does not have access to any altitude data (height maps etc) from the game so I can't automatically make an effect appear exactly at ground level. For planes, vehicles etc. the simulator places them on ground level automatically but effects will appear at the altitude specified - be it underground or high up in the air.


-- 2015-03-01 --
* Added Skin Manager, allowing user-made skins to be set on individual aircraft.
* Added Priority (1-3) to waypoints and commands.
* Added Delete Mission option to Menu. Deletes mission from internal SMB database. Any generated .Mission file is not deleted.
* Re-added precipitation controls to weather editor
* Added automatic append of backslash to bos.data.directory property if missing.
* Fixed bug resulting in Static Objects being placed at the western edge of the map.


-- 2015-02-24 --
* Added possibility to select player plane in formation. Select AI level 0. Now a new dropdown appears below "AI Level" where the player can select from 1 -> size of flight. If size of flight is changed to a value smaller than the currently selected "Player in formation" value, the latter is automatically set to the last plane of the flight.
* Added "duration" to the ATTACK_AREA command on waypoints. 1-120 minutes.
* Added checkboxes to ATTACK_AREA command for specifying which type of targets to target: Air, Ground, Vehicles. I think "Vehicles" translate to all movable ground objects while "Ground" translates to... not sure.
* Added context-aware Command editing - the relevant form components are now shown for ATTACK_AREA, ATTACK_TARGET, COVER, LAND, START and FLY commands.
* Added validation so COVER command has a target object set
* Removed controls from Weather Editor having no effect. (Note sure yet whether temperature does anything. I do think turbulence works)
* Fixed bug where all AI flights got AI level 2.
* Fixed bug where unit having no waypoints caused export to .Mission format to crash

IMPORTANT HAPPY NOTE 1: As per game version 1.09 it is no longer necessary to run user-made missions in Offline Mode
IMPORTANT HAPPY NOTE 2: Per the above, unlocks seems to work fine in user-made missions



-- 2015-02-22 --
* Added weather editor
* Fixed waypoint selection bug


-- 2015-02-20 --
* Vehicle groups command to use follow roads formation nows works
* Formation calculations for LINE, COLUMN and VEE now works
* Air units always spawn in VEE formation with > 100 meters between each aircraft. Seems to reduce collisions a lot.
* Air units try to form loose VEE formation 13 seconds after mission start (not 100% sure this actually works)
* Fixed broken map filters for Air, Ground and static objects
* Added map filter to show both sides at the same time
* Fixed bug where Select Country menu option was not visible after creating a new mission
* Vehicle formation select now only shows ground unit formations.
* Briefing icons now always shows BLUE for friendly forces and RED for enemies
* Fixed bug with wrong icons shown for various ground objects

Known/new issues:
* Aircraft groups in mission briefing uses Baloon icon. Havn't figured out the IconId for planes yet...
* All other known ones, see initial post...

src/main/webapp/js/renderer.js
/**
 * All code that actually draws something to the map canvas should go into this "class"
 */
var renderer = new function() {

    var images = {};
    var numImages = 0;

    var sources = {
        "t34": 'images/icon-t34.png',
        "pziv": 'images/pziv.png',
        "101": 'images/101.png',
        "201": 'images/201.png',
        "bf109" : 'images/bf109.png'
    };


    // get num of sources
    for(var src in sources) {
        numImages++;
    }
    for(var src in sources) {
        images[src] = new Image();
        images[src].onload = function() {

        };
        images[src].src = sources[src];
    }

    this.renderSelectionBox = function(context) {
        var box = state.getSelectionBox();
        context.beginPath();
        context.rect(box.topX, box.topY, box.bottomX, box.bottomY);
        context.fillStyle = 'red';
        context.globalAlpha = 0.1;
        context.fill();
        context.lineWidth = 1;
        context.strokeStyle = 'black';
        context.stroke();
        context.globalAlpha = 1;
    }

    this.renderAirfields = function(viewport, context) {
        if(statics.getAirfields().length == 0) {
            return;
        }

        for(var a = 0; a < statics.getAirfields().length; a++) {
            var airfield = statics.getAirfields()[a];
            var coords = coordTranslator.worldToImageInViewport(airfield.x, airfield.z, viewport, maprenderer.mapWidth, maprenderer.mapHeight);

            // Primitive occlusion culling
            if(coords.x < 0 || coords.y < 0 || coords.x > maprenderer.mapWidth || coords.y >  maprenderer.mapHeight) continue;

            drawAirfield(coords.x, coords.y, airfield.name, context);
        }
    }

    this.renderEffects = function(viewport, context) {
        if(util.isNull(state.getCurrentMission()) || util.isNull(state.getCurrentMission().effects) || state.getCurrentMission().effects.length == 0) {
            return;
        }

        for(var a = 0; a < state.getCurrentMission().effects.length; a++) {
            var effect = state.getCurrentMission().effects[a];
            var coords = coordTranslator.worldToImageInViewport(effect.x, effect.z, viewport, maprenderer.mapWidth, maprenderer.mapHeight);

            // Primitive occlusion culling
            if(coords.x < 0 || coords.y < 0 || coords.x > maprenderer.mapWidth || coords.y >  maprenderer.mapHeight) continue;

            drawEffect(coords.x, coords.y, effect.effectType, context, effect);
        }
    }

    var drawEffect = function(x, y, name, context, effect) {
        context.save();
        context.beginPath();
        context.arc(x, y, 12, 0, 2*Math.PI, false);
        context.fillStyle = 'orange';
        context.globalAlpha = 0.4;
        context.fill();
        context.globalAlpha = 0.9;
        if(util.notNull(state.getSelectedEffect()) && effect.clientId == state.getSelectedEffect().clientId) {
            context.lineWidth = 4;
        } else {
            context.lineWidth = 2;
        }
        context.strokeStyle = 'orange';
        context.stroke();
        context.globalAlpha = 1;
        context.fillStyle = 'black';
        context.font = '10pt Open Sans';
        context.textAlign = 'center';

        context.fillText(name, x, y+12+12);
        context.restore();
    }

    var drawAirfield = function(x, y, name, context) {
        context.save();
        context.beginPath();
        context.arc(x, y, 12, 0, 2*Math.PI, false);
        context.fillStyle = 'green';
        context.globalAlpha = 0.1;
        context.fill();
        context.globalAlpha = 0.5;
        context.lineWidth = 1;
        context.strokeStyle = 'green';
        context.stroke();
        context.globalAlpha = 1;
       // context.font = '12pt Open Sans';
       // context.textAlign = 'center';

      //  context.fillText(name, x, y+12+12);
        context.restore();
    }

    this.renderTriggerZones = function(metadata, zoom, viewport, context) {
        if(!util.notNull(state.getCurrentMission())) return;

         var zones = state.getCurrentMission().triggerZones;
        for(var a = 0; a < zones.length; a++) {
            var tz = zones[a];
            var coords = coordTranslator.worldToImageInViewport(tz.x, tz.z, viewport, maprenderer.mapWidth, maprenderer.mapHeight);
            drawTriggerZone(coords.x, coords.y, tz.radius/metadata.xd/zoom, context);
        }
    }

    this.renderPlacingTriggerZone = function(metadata, zoom, viewport, context) {
        if(util.notNull(state.getSelectedTriggerZone())) {
            var tz = state.getSelectedTriggerZone();
            var coords = coordTranslator.worldToImageInViewport(tz.x, tz.z, viewport, maprenderer.mapWidth, maprenderer.mapHeight);
            drawTriggerZone(coords.x, coords.y, tz.radius/metadata.xd/zoom, context);
        }

    }

    var drawTriggerZone = function(x, y, radius, context) {
        context.save();
        // Outer circle
        context.beginPath();
        context.arc(x, y, radius, 0, 2 * Math.PI, false);
        context.fillStyle = 'yellow';
        context.globalAlpha = 0.5;
        context.fill();
        context.globalAlpha = 1.0;

        context.lineWidth = 2;


        context.strokeStyle = '#444444';
        context.stroke();

        // Inner circle
        context.beginPath();
        context.arc(x, y, 16, 0, 2 * Math.PI, false);
        context.fillStyle = 'yellow';
        context.globalAlpha = 0.5;
        context.fill();
        context.globalAlpha = 1.0;

        context.lineWidth = 2;


        context.strokeStyle = '#444444';
        context.stroke();


        context.font = '12pt Open Sans';
        context.textAlign = 'center';
        context.fillStyle = 'black';
        context.fillText("TZ", x, y+6);
        context.restore();
    }

    /**
     * Renders line from last waypoint when in adding waypoint state.
     * @param mouseX
     * @param mouseY
     * @param viewport
     * @param context
     */
    this.renderPlacingWaypoint = function(mouseX, mouseY, viewport, context) {
        if(state.getState() == state.PLACING_WAYPOINT && util.notNull(state.getSelectedUnitGroup())) {
            var ug = state.getSelectedUnitGroup();
            var startingCoord = resolveStartingCoord(ug);
            context.save();
            var coord = coordTranslator.worldToImageInViewport(startingCoord.x, startingCoord.z, viewport, maprenderer.mapWidth, maprenderer.mapHeight);
            context.moveTo(coord.x, coord.y);
            context.lineWidth = 2;
            context.lineTo(mouseX, mouseY);
            context.strokeStyle = '#005500';
            context.stroke();
            context.restore();
        }
    }

    var resolveStartingCoord = function(ug) {
        if(ug.waypoints.length == 0) {
            return {x:ug.x, z:ug.z};
        } else {
            return {x:ug.waypoints[ug.waypoints.length-1].x, z:ug.waypoints[ug.waypoints.length-1].z};
        }
    }

    this.renderFlightObjects = function(viewport, context) {
        if(!(util.notNull(state.getCurrentMission()) && util.notNull(state.getCurrentMission().sides[state.getCurrentCountry()])))
            return;


        context.save();
        var unitGroups = state.getCurrentMission().sides[state.getCurrentCountry()].unitGroups;
        renderSide(state.getCurrentCountry(), viewport, context, 1.0);
        if(state.getFilter('both_sides')) {
            renderSide(state.getOtherCountry(), viewport, context, 0.3);
        }
    }

    var renderSide = function(countryId, viewport, context, globalAlpha) {
        var unitGroups = state.getCurrentMission().sides[countryId].unitGroups;
        if(unitGroups.length > 0) {
            for(var a = 0; a < unitGroups.length; a++) {
                var ug = unitGroups[a];
                if(!((ug.groupType == 'AIR_GROUP' && state.getFilter('plane_group')) || (ug.groupType == 'GROUND_GROUP' && state.getFilter('ground_group')))) {
                    continue;
                }
                var coord = coordTranslator.worldToImageInViewport(ug.x, ug.z, viewport, maprenderer.mapWidth, maprenderer.mapHeight);
                // Primitive occlusion culling
                //if(coord.x < 0 || coord.y < 0 || coord.x > maprenderer.mapWidth || coord.y >  maprenderer.mapHeight) continue;

                drawUnitGroup(coord.x, coord.y, ug, context, globalAlpha);

                if(countryId == state.getCurrentCountry()) {
                    for(var b = 0; b < ug.waypoints.length; b++) {
                        var waypoint = ug.waypoints[b];
                        var wpCoord = coordTranslator.worldToImageInViewport(waypoint.x, waypoint.z, viewport, maprenderer.mapWidth, maprenderer.mapHeight);
                        var lastCoord = {};
                        if(b > 0) {
                            lastCoord = coordTranslator.worldToImageInViewport(ug.waypoints[b-1].x, ug.waypoints[b-1].z, viewport, maprenderer.mapWidth, maprenderer.mapHeight);
                        }
                        var isSelectedWaypoint = waypoint == state.getSelectedWaypoint();
                        drawWaypoint(wpCoord.x, wpCoord.y, lastCoord.x, lastCoord.y, coord.x, coord.y, b, isSelectedWaypoint, ug.clientId, waypoint, context);
                    }
                }
            }
        }

        var staticObjectGroups = state.getCurrentMission().sides[countryId].staticObjectGroups;
        if(staticObjectGroups.length > 0 && state.getFilter('static_object')) {
            for(var a = 0; a < staticObjectGroups.length; a++) {
                var sog = staticObjectGroups[a];
                var coord = coordTranslator.worldToImageInViewport(sog.x, sog.z, viewport, maprenderer.mapWidth, maprenderer.mapHeight);
                drawStaticObjectGroup(coord.x, coord.y, sog, context, globalAlpha);
            }
        }
        context.restore();
    }

    var drawWaypoint = function(x, y, x2, y2, ugx, ugy, index, isSelectedWaypoint, unitGroupClientId, waypoint, context) {
        var alpha = util.notNull(state.getSelectedUnitGroup()) && unitGroupClientId == state.getSelectedUnitGroup().clientId ? 0.9 : 0.2;
        context.globalAlpha = alpha;
        if(index == 0) {
            // First, draw line from parent object.
            context.moveTo(ugx, ugy);

            context.lineWidth = 2;
            context.lineTo(x, y);
            context.strokeStyle = '#550000';
            context.stroke();
        } else {
            // Draw line from previous to here
            context.moveTo(x2, y2);

            context.lineWidth = 2;
            context.lineTo(x, y);
            context.strokeStyle = '#550000';
            context.stroke();
        }

        context.beginPath();
        context.arc(x, y, 8, 0, 2 * Math.PI, false);
        context.fillStyle = 'red';
        context.globalAlpha = alpha;
        context.fill();
        context.globalAlpha = alpha;
        if(isSelectedWaypoint) {
            context.lineWidth = 4;
        } else {
             context.lineWidth = 2;
        }

        context.strokeStyle = '#550000';
        context.stroke();


        context.font = '10pt Open Sans';
        context.textAlign = 'center';
        context.fillStyle = 'white';
        context.fillText(""+ (index+1), x, y+4);

        // For waypoints with a non-fly command, print it..
        if(util.notNull(waypoint.action) && util.notNull(waypoint.action.actionType) && waypoint.action.actionType != 'FLY') {
            context.font = '8pt Open Sans';
            context.textAlign = 'center';
            context.fillStyle = 'black';
            context.fillText(waypoint.action.actionType, x, y+22);
        }

    }

    var drawStaticObjectGroup  = function(x, y, staticObjectGroup, context, globalAlpha) {
        context.beginPath();
        context.arc(x, y, 20, 0, 2 * Math.PI, false);
        context.fillStyle = 'purple';
        context.globalAlpha = globalAlpha-0.2;
        context.fill();
        context.globalAlpha = globalAlpha;
        if(util.notNull(state.getSelectedStaticObjectGroup()) && staticObjectGroup.clientId == state.getSelectedStaticObjectGroup().clientId) {
            context.lineWidth = 6;
        } else {
            context.lineWidth = 3;
        }

        context.strokeStyle = '#222222';
        context.stroke();

        context.font = '12pt Open Sans';
        context.textAlign = 'center';

        context.fillText(staticObjectGroup.type, x, y+20+16);

        context.font = '10pt Open Sans';
        context.textAlign = 'center';

        context.fillStyle = 'white';
        context.fillText("S", x, y+4);
    }

    var TO_RADIANS = Math.PI/180;

    var drawUnitGroup = function(x, y, unitGroup, context, globalAlpha) {

        var type = unitGroup.groupType;

        context.beginPath();
        context.arc(x, y, 20, 0, 2 * Math.PI, false);
        context.fillStyle = type == 'AIR_GROUP' ? 'blue' : 'green';

        if(type == 'AIR_GROUP') {
            context.globalAlpha = globalAlpha == 1.0 ? 0.2 : 0.1;
        } else {
            context.globalAlpha = globalAlpha == 1.0 ? 0.7 : 0.3;
        }
        context.fill();
        context.globalAlpha = globalAlpha;
        if(util.notNull(state.getSelectedUnitGroup()) && unitGroup.clientId == state.getSelectedUnitGroup().clientId) {
            context.lineWidth = 5;
        } else {
            context.lineWidth = 3;
        }

        context.strokeStyle = globalAlpha == 1.0 ? '#000055' : '#FF0000';
        context.stroke();

        context.font = '12pt Open Sans';
        context.textAlign = 'center';
        context.fillStyle = 'black';
        context.fillText(unitGroup.type, x, y+20+16);


        context.font = '10pt Open Sans';
        context.fillText(unitGroup.size, x-18, y-18);

       // var imageObj = new Image();
        var scale = 0.5;
        var rotation = 0;

        var src = '';
        if(type == 'AIR_GROUP') {
            rotation = Math.round( unitGroup.yOri );
            src = 'bf109';
        } else {
            src = state.getCurrentCountry() == 201 ? "pziv" : "t34";
            scale = 0.6;
        }
        context.save();
        context.translate(x,y);
        context.rotate(rotation * TO_RADIANS);
        context.scale(scale, scale);
        context.drawImage(images[src], -images[src].width/2, -images[src].height/2);
        context.restore();
    }

    this.renderKillsOnMapVectorized = function(viewport, context, data) {

        context.save();
        if(data.kills.length > 0) {
            for(var a = 0; a < data.kills.length; a++) {
                var kill = data.kills[a];
                if(kill.parentId == -1 && kill.killedXPos != null && kill.killedZPos != null)   {
                    var coord = coordTranslator.worldToImageInViewport(kill.killedXPos, kill.killedZPos, viewport, maprenderer.mapWidth, maprenderer.mapHeight);
                    drawKill(coord.x, coord.y, 10, kill.type, context);
                }
            }
        }
        context.restore();
    }

    this.renderFlightMapVectorized = function(viewport, context, data) {

        context.save();
        if(data.flightTrack.length > 1) {
            var startingCoord = coordTranslator.worldToImageInViewport(data.flightTrack[0].x, data.flightTrack[0].z, viewport, maprenderer.mapWidth, maprenderer.mapHeight);

            context.beginPath();

            context.moveTo(startingCoord.x, startingCoord.y);
            for(var a = 1; a < data.flightTrack.length; a++) {
                var fp = data.flightTrack[a];
                drawFlightPathVectorized(viewport, fp, context);
            }
            context.stroke();
        }
        context.moveTo(startingCoord.x, startingCoord.y);
        drawStartSymbol(startingCoord.x, startingCoord.y, context);
        var secondEndingCoord = coordTranslator.worldToImageInViewport(data.flightTrack[data.flightTrack.length - 2].x, data.flightTrack[data.flightTrack.length - 2].z, viewport, maprenderer.mapWidth, maprenderer.mapHeight);
        var endingCoord = coordTranslator.worldToImageInViewport(data.flightTrack[data.flightTrack.length - 1].x, data.flightTrack[data.flightTrack.length - 1].z, viewport, maprenderer.mapWidth, maprenderer.mapHeight);
        drawEndSymbol(endingCoord.x, endingCoord.y, secondEndingCoord.x, secondEndingCoord.y, context);
        context.restore();
    }


    this.renderKillInfoDialog = function(kill) {
        $('#gid_objectname').text(kill.type);
        $('#gid_timekilled').text(kill.gameTime);     gid_side
        $('#gid_side').text(kill.country);
        $('#gid_hitcount').text(kill.totalHits);

        var hitsByType = kill.hits;
        $('#gid_hitsbytype').empty();
        if(hitsByType != null) {
            for (var key in hitsByType) {
                if (hitsByType.hasOwnProperty(key)) {
                    var tpl = '<div class="row">' +
                        '<div class="col-md-10">' + key + '</div>' +
                        '<div class="col-md-2">' + hitsByType[key] + '</div>' +
                        '</div>';
                    $('#gid_hitsbytype').append(tpl);
                }
            }
        }
        $('#gid-dialog').removeClass("hidden");
    }

     var drawFlightPathVectorized = function(viewport, fp, context) {
        var coord = coordTranslator.worldToImageInViewport(fp.x, fp.z, viewport, maprenderer.mapWidth, maprenderer.mapHeight);
        context.lineWidth = 2;
        context.lineTo(coord.x, coord.y);
        context.arc(coord.x, coord.y, 2, 0, 2*Math.PI, false);
        context.moveTo(coord.x, coord.y);
    }

    var drawKill = function(x, y, size, label, context) {

        context.beginPath();
        context.arc(x, y, size, 0, 2 * Math.PI, false);
        context.fillStyle = 'red';
        context.globalAlpha = 0.8;
        context.fill();
        context.globalAlpha = 1.0;
        context.lineWidth = 2;
        context.strokeStyle = '#550000';
        context.stroke();

        context.font = '12pt Open Sans';
        context.textAlign = 'center';

        context.fillText(label, x, y+size+16);

        context.font = '10pt Open Sans';
        context.textAlign = 'center';

        context.fillStyle = 'white';
        context.fillText("K", x, y+4);

    }

    var drawStartSymbol = function(x, y, context) {
        context.save();

        context.beginPath();
        context.arc(x, y, 15, 0, Math.PI*2, false);
        context.closePath();
        context.lineWidth = 3;
        context.fillStyle = 'green';
        context.globalAlpha = 0.8;
        context.fill();
        context.globalAlpha = 1.0;
        context.strokeStyle = '#005500';
        context.stroke();

        context.font = '12pt arial';
        context.textAlign = 'center';

        context.fillStyle = 'white';
        context.fillText("S", x, y+5);

        context.restore();
    }

    var drawEndSymbol = function(x, y, x2, y2, context) {
        context.save();
        context.beginPath();
        context.moveTo(x2, y2);
        context.lineWidth = 2;
        context.lineTo(x, y);
        context.closePath();
        context.stroke();

        context.beginPath();
        context.arc(x, y, 15, 0, Math.PI*2, false);
        context.closePath();
        context.lineWidth = 3;
        context.fillStyle = 'red';
        context.globalAlpha = 0.8;
        context.fill();
        context.globalAlpha = 1.0;
        context.strokeStyle = '#550000';
        context.stroke();

        context.font = '12pt arial';
        context.textAlign = 'center';

        context.fillStyle = 'white';
        context.fillText("E", x, y+5);
        context.restore();
    }
}